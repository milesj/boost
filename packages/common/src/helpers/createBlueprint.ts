import { Blueprint, schemas } from 'optimal';
import { BlueprintFactory } from '../types';

/**
 * Can be used to generate a blueprint object for use within
 * [optimal](https://github.com/milesj/optimal) checks. All supported optimal
 * schemas are passed as an object to the factory.
 *
 * ```ts
 * import { createBlueprint } from '@boost/common';
 * import { optimal } from '@boost/common/optimal';
 *
 * const blueprint = createBlueprint(({ string, number }) => ({
 * 	name: string().required(),
 * 	age: number().gt(0),
 * }));
 *
 * const object = optimal(blueprint).validate({});
 * ```
 */
export function createBlueprint<T extends object>(factory: BlueprintFactory<T>): Blueprint<T> {
	return factory(schemas);
}
