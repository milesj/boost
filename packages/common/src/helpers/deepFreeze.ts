import isObject from './isObject';

export default function deepFreeze<T extends object = object>(obj: T): T {
  if (Object.isFrozen(obj)) {
    return obj as T;
  }

  const nextObj: Record<string, unknown> = {};

  Object.entries(obj).forEach(([key, value]) => {
    // Only freeze plain objects
    if (isObject(value) && value.constructor === Object) {
      nextObj[key] = deepFreeze(value);
    } else {
      nextObj[key] = value;
    }
  });

  return Object.freeze(nextObj) as T;
}
