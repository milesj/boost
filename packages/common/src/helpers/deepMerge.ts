import { isObject } from './isObject';

export type MergableArray = unknown[];
export type MergableObject = Record<string, unknown>;
export type Mergeable = MergableArray | MergableObject;
export type InferMergeable<T> = T extends unknown[]
	? MergableArray
	: T extends object
		? MergableObject
		: never;

function merge<T extends Mergeable>(prev: T, next: unknown): T {
	const base = prev as MergableObject;

	Object.entries(next as Mergeable).forEach(([key, value]) => {
		const prevValue = base[key];

		if (isObject(prevValue) && isObject(value)) {
			base[key] = merge({ ...prevValue }, value);
		} else if (Array.isArray(prevValue) && Array.isArray(value)) {
			base[key] = merge([...(prevValue as MergableArray)], value);
		} else {
			base[key] = value;
		}
	});

	return base as T;
}

/**
 * Can be used to recursively merge objects and arrays, where values on the
 * right-hand side will overwrite values on the left-hand based on the key
 * or index respectively. Furthermore, if the 2nd argument is not provided,
 * it will simply clone the base value.
 *
 * ```ts
 * import { deepMerge } from '@boost/common';
 *
 * const obj = deepMerge({ foo: 123, bar: 'abc' }, { foo: 456, baz: true });
 *
 * // -> { foo: 456, bar: 'abc', baz: true }
 * ```
 */
export function deepMerge<T extends Mergeable, V extends InferMergeable<T>>(base: T, other?: V): T {
	const next = Array.isArray(base)
		? merge<MergableArray>([], base)
		: merge<MergableObject>({}, base);

	if (other) {
		merge(next, other);
	}

	return next as T;
}
