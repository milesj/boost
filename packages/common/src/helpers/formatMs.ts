import prettyMs, { Options } from 'pretty-ms';

export default function formatMs(ms: number, options?: Options): string {
  if (!Number.isFinite(ms) || ms === 0) {
    return '0s';
  }

  return prettyMs(ms, { keepDecimalsOnWholeSeconds: true, ...options });
}
