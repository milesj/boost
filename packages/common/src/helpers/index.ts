export * from './createBlueprint';
export * from './deepFreeze';
export * from './deepMerge';
export * from './formatMs';
export * from './instanceOf';
export * from './isEmpty';
export * from './isFilePath';
export * from './isModuleName';
export * from './isObject';
export * from './isPlainObject';
export * from './toArray';
