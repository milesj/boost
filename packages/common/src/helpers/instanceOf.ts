import { Constructor } from '../types';

/**
 * Native `instanceof` checks are problematic, as cross realm checks fail.
 * They will also fail when comparing against source and built files.
 * So emulate an `instanceof` check by comparing constructor names.
 */
export default function instanceOf<T = unknown>(
  object: unknown,
  declaration: Constructor<T>,
  loose: boolean = true,
): object is T {
  if (!object || typeof object !== 'object') {
    return false;
  }

  if (object instanceof declaration) {
    return true;
  }

  if (!loose) {
    return false;
  }

  let current = object;

  while (current) {
    if (current.constructor.name === 'Object') {
      break;
    }

    if (
      current.constructor.name === declaration.name ||
      // istanbul ignore next
      (current instanceof Error && current.name === declaration.name)
    ) {
      return true;
    }

    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    current = Object.getPrototypeOf(current);
  }

  return false;
}
