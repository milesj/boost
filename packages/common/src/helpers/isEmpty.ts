import { isObject } from './isObject';

export function isEmpty(value: unknown): boolean {
	return (
		!value ||
		(Array.isArray(value) && value.length === 0) ||
		(isObject<object>(value) && Object.keys(value).length === 0) ||
		false
	);
}
