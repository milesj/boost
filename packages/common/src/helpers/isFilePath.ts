import { Path } from '../Path';
import { PortablePath } from '../types';

const NIX_START = /^(\/|\.)/u;
const WIN_START = /^([A-Z]:|\.)/u;

export function isFilePath(path: PortablePath): boolean {
	const filePath = path instanceof Path ? path.path() : path;

	if (filePath === '') {
		return false;
	}

	// istanbul ignore next
	if (process.platform === 'win32') {
		return WIN_START.test(filePath) || filePath.includes('/') || filePath.includes('\\');
	}

	return NIX_START.test(filePath) || filePath.includes('/');
}
