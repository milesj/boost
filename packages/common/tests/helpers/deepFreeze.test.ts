import { describe, expect,it } from 'vitest';
import { deepFreeze } from '../../src/helpers/deepFreeze';

describe('deepFreeze()', () => {
	it('returns a frozen object', () => {
		const obj = deepFreeze({
			foo: {
				bar: {
					baz: 123,
				},
			},
		});

		expect(obj).toEqual({
			foo: {
				bar: {
					baz: 123,
				},
			},
		});

		expect(() => {
			// @ts-expect-error Allow type
			obj.foo.bar = 123;
		}).toThrowErrorMatchingSnapshot();
	});

	it('returns already frozen objects', () => {
		const obj1 = Object.freeze({ foo: 123 });
		const obj2 = deepFreeze(obj1);

		expect(obj1).toBe(obj2);
	});
});
