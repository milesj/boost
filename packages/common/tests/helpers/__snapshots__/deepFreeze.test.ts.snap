// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deepFreeze() returns a frozen object 1`] = `"Cannot assign to read only property 'bar' of object '#<Object>'"`;
